src/main/java/io/odpf/firehose/sinkdecorator/SinkWithDlq.java

```diff
@@ -5,97 +5,79 @@
-    private final String topic;
-    private BackOffProvider backOffProvider;
+    public static final String DLQ_BATCH_KEY = "dlq-batch-key";
+    private final DlqWriter writer;
+    private final BackOffProvider backOffProvider;
+    private final int maxRetryAttempts;
+    private final boolean isFailOnMaxRetryAttemptsExceeded;
+
     private Instrumentation instrumentation;
 
-    public SinkWithDlq(Sink sink, Producer<byte[], byte[]> kafkaProducer, String topic,
-                       Instrumentation instrumentation, BackOffProvider backOffProvider) {
+    public SinkWithDlq(Sink sink, DlqWriter writer, BackOffProvider backOffProvider, int maxRetryAttempts, boolean isFailOnMaxRetryAttemptsExceeded, Instrumentation instrumentation) {
         super(sink);
-        this.kafkaProducer = kafkaProducer;
-        this.topic = topic;
+        this.writer = writer;
         this.backOffProvider = backOffProvider;
+        this.maxRetryAttempts = maxRetryAttempts;
+        this.isFailOnMaxRetryAttemptsExceeded = isFailOnMaxRetryAttemptsExceeded;
         this.instrumentation = instrumentation;
     }
 
-    public static SinkWithDlq withInstrumentationFactory(Sink sink, Producer<byte[], byte[]> kafkaProducer, String topic,
-                                                         StatsDReporter statsDReporter, BackOffProvider backOffProvider) {
-        return new SinkWithDlq(sink, kafkaProducer, topic, new Instrumentation(statsDReporter, SinkWithDlq.class), backOffProvider);
+    public static SinkWithDlq withInstrumentationFactory(Sink sink, DlqWriter dlqWriter, BackOffProvider backOffProvider, int maxRetryAttempts, boolean isFailOnMaxRetryAttemptsExceeded, StatsDReporter statsDReporter) {
+        return new SinkWithDlq(sink, dlqWriter, backOffProvider, maxRetryAttempts, isFailOnMaxRetryAttemptsExceeded, new Instrumentation(statsDReporter, SinkWithDlq.class));
     }
 
     /**
      * Pushes all the failed messages to kafka as DLQ.
      *
-     * @param message list of messages to push
+     * @param messages list of messages to push
      * @return list of failed messaged
      * @throws IOException
      * @throws DeserializerException
      */
     @Override
-    public List<Message> pushMessage(List<Message> message) throws IOException, DeserializerException {
-        List<Message> retryQueueMessages = super.pushMessage(message);
-        int attemptCount = 0;
-
-        while (!retryQueueMessages.isEmpty()) {
-            instrumentation.captureRetryAttempts();
-            retryQueueMessages = pushToKafka(retryQueueMessages);
-            backOffProvider.backOff(attemptCount++);
+    public List<Message> pushMessage(List<Message> messages) throws IOException, DeserializerException {
+        List<Message> retryQueueMessages = super.pushMessage(messages);
+        if (super.canManageOffsets()) {
+            super.addOffsets(DLQ_BATCH_KEY, retryQueueMessages);
+            retryQueueMessages = pushToWriter(retryQueueMessages);
```

Comment:
let's push the messages first then call addOffsets(). Also this looks wrong because here not all the messages might not be pushed, but we are setting them to committable.

Reply:
Also we might need to check message errorinfo here.. the return value of this method call will not be used anywhere. So here we need to throw the exception if not all messages were pushed to DLQ.

Reply:
there is an exception already 

```
  @Key("DLQ_FAIL_ON_MAX_RETRY_ATTEMPTS_EXCEEDED")
    @DefaultValue("true")
    boolean getFailOnMaxRetryAttemptsExceeded();
```
 
```
   if (!retryQueueMessages.isEmpty() && isFailOnMaxRetryAttemptsExceeded) {
            throw new IOException("exhausted maximum number of allowed retry attempts to write messages to DLQ");
        }
```

Reply:
yes agree only offset that already pushed that need to be set as committable, will change this